caikit_nlp/toolkit/text_generation/model_run_utils.py

```diff
@@ -0,0 +1,241 @@
+    error.type_check("<NLP65883534E>", str, text=text)
+    error.type_check(
+        "<NLP03860680E>", int, allow_none=True, max_new_tokens=max_new_tokens
+    )
+    error.type_check(
+        "<NLP30091276E>", int, allow_none=True, min_new_tokens=min_new_tokens
+    )
+    error.type_check(
+        "<NLP55411551E>",
+        int,
+        allow_none=True,
+        truncate_input_tokens=truncate_input_tokens,
+    )
+    error.type_check("<NLP84635843E>", int, allow_none=True, top_k=top_k)
+    error.type_check("<NLP55267523E>", float, allow_none=True, top_p=top_p)
+    error.type_check("<NLP13670202E>", float, allow_none=True, typical_p=typical_p)
+    error.type_check(
+        "<NLP11929418E>", float, allow_none=True, repetition_penalty=repetition_penalty
+    )
+    error.type_check_all(
+        "<NLP41311583E>", str, allow_none=True, stop_sequences=stop_sequences
+    )
+
+    # NOTE: below is to match TGIS API, where 0 identifies as no truncation
+    if truncate_input_tokens == 0:
+        # NOTE: below will make model throw error in case inputs are longer
+        # than allowed length
+        truncation = False
+
+    else:
+        truncation = True
+
+    if repetition_penalty == 0.0:
+        repetition_penalty = 1.0
+
+    gen_optional_params = {}
+
+    # TODO: Make decoding parameters enums
+    if decoding_method == "SAMPLING":
+        gen_optional_params["do_sample"] = True
+        gen_optional_params["top_k"] = top_k
+        gen_optional_params["top_p"] = top_p
+        gen_optional_params["typical_p"] = typical_p
+        gen_optional_params["temperature"] = temperature
+
+    if stop_sequences and len(stop_sequences) > 0:
+        # Tokenize sequences
+        stop_sequence_ids = tokenizer.encode(stop_sequences)
+        stopping_criteria = SequenceStoppingCriteria(stop_sequence_ids)
+        gen_optional_params["stopping_criteria"] = stopping_criteria
+
+    tok_tensors = tokenizer(
+        text,
+        truncation=truncation,
+        max_length=truncate_input_tokens,
+        return_tensors="pt",
+    )
+    inputs = {k: v.to(model.device) for k, v in tok_tensors.items()}
+    with torch.no_grad():
+        generate_ids = model.generate(
+            input_ids=inputs["input_ids"],
+            max_new_tokens=max_new_tokens,
+            min_new_tokens=min_new_tokens,
+            repetition_penalty=repetition_penalty,
+            use_cache=True,
+            max_time=max_time,
+            exponential_decay_length_penalty=exponential_decay_length_penalty,
+            **gen_optional_params,
+            **kwargs,
+        )
+
+    token_count = generate_ids.size(1) - 1
+
+    preds = [
+        tokenizer.decode(g, skip_special_tokens=True, clean_up_tokenization_spaces=True)
+        for g in generate_ids
+    ]
+
+    if generate_ids[0][-1].item() == eos_token:
+        finish_reason = "EOS_TOKEN"
+    elif generate_ids.size(1) - 1 == max_new_tokens:
+        finish_reason = "MAX_TOKENS"
```

Comment:
Definitely a technicality, but I'm wondering if MAX_TOKENS should be the reason if we hit the maximum number of tokens and the last one is an EOS 🤔 in my opinion, this makes a little more sense

Reply:
hmm. just to make sure I am understanding correctly. I guess, that would depend if the last token would still be EOS in case we hit the size 🤔

Reply:
No idea! I think I can check that and get back to you 😄

Let's not let that block this PR though, can submit a patch if it's not aligned later, since this is a super obscure case and both make sense